File: app/src/main/java/se/jakob/knarkklocka/data/AlarmDatabase.java
package se.jakob.knarkklocka.data;

import android.arch.persistence.db.SupportSQLiteDatabase;
import android.arch.persistence.room.Database;
import android.arch.persistence.room.Room;
import android.arch.persistence.room.RoomDatabase;
import android.arch.persistence.room.TypeConverters;
import android.content.Context;
import android.os.AsyncTask;
import android.support.annotation.NonNull;
import android.util.Log;

import java.util.Date;

import se.jakob.knarkklocka.BuildConfig;

@TypeConverters(DateConverter.class)
@Database(entities = {Alarm.class}, version = 1, exportSchema = false)
public abstract class AlarmDatabase extends RoomDatabase {
    private static final String LOG_TAG = AlarmDatabase.class.getSimpleName();
    private static final Object LOCK = new Object();
    private static volatile AlarmDatabase INSTANCE; //Instance of the database

    public abstract AlarmDao alarmDao(); //Getter for Dao, use to access the database

    /*For debug purposes, populates the database with alarms*/
    private static RoomDatabase.Callback sRoomDatabaseCallback =
            new RoomDatabase.Callback() {
                @Override
                public void onOpen(@NonNull SupportSQLiteDatabase db) {
                    super.onOpen(db);
                    new PopulateDbAsync(INSTANCE).execute();
                }
            };

    public static AlarmDatabase getDatabase(final Context context) {
        if (INSTANCE == null) {
            synchronized (LOCK) {
                if (INSTANCE == null) {
                    Log.d(LOG_TAG, "Creating new database instance");
                    if (BuildConfig.DEBUG) {
                        INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                                AlarmDatabase.class, "alarm_database_debug")
                                .build();
                    } else {
                        INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                                AlarmDatabase.class, "alarm_database")
                                .build();
                    }
                }
            }
        }
        return INSTANCE;
    }

    /*For debug purposes, populates the database with alarms*/
    private static class PopulateDbAsync extends AsyncTask<Void, Void, Void> {
        private final AlarmDao mDao;
        PopulateDbAsync(AlarmDatabase db) {
            mDao = db.alarmDao();
        }
        @Override
        protected Void doInBackground(final Void... params) {
            mDao.deleteAll();
            Alarm alarm;
            alarm = new Alarm(Alarm.STATE_DEAD, new Date(500), new Date(500));
            mDao.insert(alarm);
            alarm = new Alarm(Alarm.STATE_DEAD, new Date(600), new Date(600));
            mDao.insert(alarm);
            return null;
        }
    }
}

File: app/src/main/java/se/jakob/knarkklocka/data/AlarmDatabase.kt
package se.jakob.knarkklocka.data


import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import androidx.sqlite.db.SupportSQLiteDatabase
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkManager
import se.jakob.knarkklocka.BuildConfig
import se.jakob.knarkklocka.workers.PopulateDatabaseWorker

@TypeConverters(Converters::class)
@Database(entities = [Alarm::class], version = 4, exportSchema = false)
abstract class AlarmDatabase : RoomDatabase() {

    abstract fun alarmDao(): AlarmDao  //Getter for Dao, use to access the database

    companion object {

        @Volatile private var instance: AlarmDatabase? = null //Instance of the database

        fun getInstance(context: Context): AlarmDatabase {
            return instance ?: synchronized(this) {
                instance ?: buildDatabase(context).also { instance = it }
            }
        }

        /*For debug purposes, populates the database with alarms*/
        private val databaseDebugCallback = object : RoomDatabase.Callback() {
            override fun onOpen(db: SupportSQLiteDatabase) {
                super.onOpen(db)
                val request = OneTimeWorkRequestBuilder<PopulateDatabaseWorker>().build()
                WorkManager.getInstance().enqueue(request)
            }
        }

        private fun buildDatabase(context: Context): AlarmDatabase {
            if (BuildConfig.DEBUG) {
                return Room.databaseBuilder(context, AlarmDatabase::class.java,
                        "alarm_database_debug")
                        .addCallback(databaseDebugCallback)
                        .fallbackToDestructiveMigration()
                        .build()
            } else {
                return Room.databaseBuilder(context,
                        AlarmDatabase::class.java, "alarm_database")
                        .fallbackToDestructiveMigration()
                        .build()
            }
        }

    }
}
